drivers/SmartThings/philips-hue/src/init.lua

```diff
@@ -982,7 +901,89 @@ cosock.spawn(function()
+                "Error Number %s in get_light_by_id response while onboarding bridge %s: %s"
+                , idx
+                , (msg_device.label or msg_device.id or "unknown device")
+                , st_utils.stringify_table(resource_err)
+              ))
+            end
+            return
+          end
+
+          for _, light in ipairs(light_resource.data or {}) do
+            local light_resource_description = {
+              id = light.id,
+              on = light.on,
+              color = light.color,
+              dimming = light.dimming,
+              color_temp = light.color_temperature,
+              mode = light.mode,
+              parent_device_id = bridge_device_uuid,
+              hue_device_id = light.owner.rid,
+              hue_device_data = device_data
+            }
+            Discovery.light_state_disco_cache[light.id] = light_resource_description
+          end
+
+          for stray_dni, stray_light in pairs(stray_lights) do
+            local matching_v1_id = stray_light.data and stray_light.data.bulbId and
+              stray_light.data.bulbId == device_data.id_v1:gsub("/lights/", "")
+            local matching_uuid = stray_light.device_network_id == svc_info.rid or
+              stray_light.device_network_id == svc_info.rid
+
+            if matching_v1_id or matching_uuid then
+              local api_key_extracted = api_instance.headers["hue-application-key"]
+              log.info_with({ hub_logs = true }, " ", stray_light.label, ", re-adding")
+              log.info_with({ hub_logs = true }, string.format(
+                'Found Bridge for stray light %s, retrying onboarding flow.\n' ..
+                '\tMatching v1 id? %s\n' ..
+                '\tMatching uuid? %s\n' ..
+                '\tlight_device DNI: %s\n' ..
+                '\tlight_device Parent Assigned Key: %s\n' ..
+                '\tlight_device parent device id: %s\n' ..
+                '\tProvided bridge_device_id: %s\n' ..
+                '\tCached API key for given bridge_device_id: %s\n' ..
+                '\tCached bridge device for given API key: %s\n'
+                ,
+                stray_light.label,
+                matching_v1_id,
+                matching_uuid,
+                stray_light.device_network_id,
+                stray_light.parent_assigned_child_key,
+                stray_light.parent_device_id,
+                bridge_device_uuid,
+                Discovery.api_keys[hue_driver:get_device_info(bridge_device_uuid).device_network_id],
+                hue_driver.api_key_to_bridge_id[api_key_extracted]
+              ))
+              stray_dni_to_rid[stray_dni] = svc_info.rid
+            end
+          end
+        end,
+        "[process_strays]"
+      )
+      log.info_with({ hub_logs = true }, "Finished querying bridge for devices from stray light handler")
+      process_strays(thread_local_driver, stray_lights, msg_device.id)
```

Comment:
Is this a recursive call?  Wouldn't the loop handle this?

Reply:
So if I understand the changes correct, what happens is that in the main coroutine/task, after a Bridge is found, all the devices are iterated and once a stray is matched to a `rid`, the following gets set `              stray_dni_to_rid[stray_dni] = svc_info.rid` which then allows the `process_strays` to then migrate the newly matched strays.

Reply:
@varzac @kianooshkarami `process_strays` isn't recursive, but it's confusing because the function is declared inside of the coroutine so that it can create a closure over variables on the closure's stack if needed. The need for this has come and gone throughout refactors.

Currently it's closing over the `stray_dni_to_rid` map. If we can get (and keep) it to be referrentially pure then we should definitely move the definition to the module local scope to reduce this confusion cuz I have confused myself with it multiple times.

Reply:
Yeah, understanding scope is just pretty hard on github.  Once I pulled this locally, this made sense.